ocaml/xapi/xapi_pvs_farm.ml

```diff
@@ -26,33 +26,108 @@ let introduce ~__context ~name =
+
+(** [sr_hosts] returns all hosts attached to a list of [sr]s. *)
+let sr_hosts ~__context srs =
+	srs
+	|> List.map (fun sr  -> Db.SR.get_PBDs ~__context ~self:sr)
+	|> List.concat
+	|> List.map (fun pbd -> Db.PBD.get_host ~__context ~self:pbd)
+
+(** [clients] returns all hosts that use the farm as a client *)
+let clients ~__context ~self =
+	proxies ~__context ~self
+	|> List.map (fun proxy -> Db.PVS_proxy.get_VIF ~__context ~self:proxy)
+	|> List.map (fun vif -> Db.VIF.get_VM ~__context ~self:vif)
+	|> List.map (fun vm  -> Db.VM.get_resident_on ~__context ~self:vm)
+
+(** add a shared SR *)
+let add_shared_cache_storage ~__context ~self ~value =
+	let cache = Db.PVS_farm.get_cache_storage ~__context ~self in
+	match cache with
+	| []    -> Db.PVS_farm.add_cache_storage ~__context ~self ~value
+	| [_]   -> not_allowed "can't add a shared SR to an existing SR"
+	| _     -> not_allowed "can't add a shared SR to a existing SRs"
+
+(* add a local SR *)
+let add_local_cache_storage ~__context ~self ~value =
+	let cache = Db.PVS_farm.get_cache_storage ~__context ~self in
+	if shared ~__context cache then
+		not_allowed "can't add SR to a cache using shared storage"
+	else if not @@ disjoint
+			(sr_hosts ~__context cache) (sr_hosts ~__context [value]) then
+		not_allowed "SR's host is already providing storage"
+	else
+		Db.PVS_farm.add_cache_storage ~__context ~self ~value
 
 let add_cache_storage ~__context ~self ~value =
-  not_implemented "PVS_farm.add_cache_storage"
+	match Db.SR.get_shared ~__context ~self:value with
+	| true  -> add_shared_cache_storage ~__context ~self ~value
+	| false -> add_local_cache_storage  ~__context ~self ~value
 
-let remove_cache_storage ~__context ~self ~value =
-  not_implemented "PVS_farm.remove_cache_storage"
+(** remove a shared SR. This should be the only one *)
+let remove_shared_cache_storage ~__context ~self ~value =
+	let cache = Db.PVS_farm.get_cache_storage ~__context ~self in
+	if not @@ List.mem value cache then
+		not_allowed "SR is not part of the farm's storage"
+	else if proxies ~__context ~self <> [] then
+		not_allowed "can't remove SR while proxies are still active"
+	else
+		assert (cache = [value]);
+		Db.PVS_farm.remove_cache_storage ~__context ~self ~value
 
+(** remove a local SR *)
+let remove_local_cache_storage ~__context ~self ~value =
+	let cache = Db.PVS_farm.get_cache_storage ~__context ~self in
+	if not @@ List.mem value cache then
+		not_allowed "SR is not part of the farm's storage"
+	else if not @@ disjoint
+			(clients ~__context ~self) (sr_hosts ~__context [value]) then
+		not_allowed "SR's host is used by the farm - can't remove it"
+	else
+		Db.PVS_farm.remove_cache_storage ~__context ~self ~value
+
+let remove_cache_storage ~__context ~self ~value =
+	match Db.SR.get_shared ~__context ~self:value with
+	| true  -> remove_shared_cache_storage ~__context ~self ~value
+	| false -> remove_local_cache_storage  ~__context ~self ~value
```

Comment:
Operationally it's the same, but I like here a case statement because we could also branch out based on other SR properties (in the future): memory SRs or whatever it is.